src/main/java/org/jitsi/videobridge/EndpointMessageTransport.java

```diff
@@ -169,106 +169,49 @@ private void sendMessage(
-        Object o = jsonObject.get("pinnedEndpoints");
-        if (!(o instanceof JSONArray))
-        {
-            logger.warn("Received invalid or unexpected JSON: " + jsonObject);
-            return;
-        }
-
-        JSONArray jsonArray = (JSONArray) o;
-        Set<String> newPinnedEndpoints = new HashSet<>();
-        for (Object endpointId : jsonArray)
-        {
-            if (endpointId != null && endpointId instanceof String)
-            {
-                newPinnedEndpoints.add((String)endpointId);
-            }
-        }
-
-        if (logger.isDebugEnabled())
-        {
-            logger.debug("Pinned " + newPinnedEndpoints);
-        }
         endpoint.pinnedEndpointsChanged(newPinnedEndpoints);
+        propagateJSONObject(jsonObject);
     }
 
     /**
      * {@inheritDoc}
      */
     @Override
-    protected void onSelectedEndpointChangedEvent(
-        Object src,
-        JSONObject jsonObject)
+    protected void onSelectedEndpointsChangedEvent(
+        JSONObject jsonObject, Set<String> newSelectedEndpoints)
     {
-        // Find the new pinned endpoint.
-        String newSelectedEndpointID
-                = (String) jsonObject.get("selectedEndpoint");
-
-        Set<String> newSelectedIDs = Collections.EMPTY_SET;
-        if (newSelectedEndpointID != null && !"".equals(newSelectedEndpointID))
-        {
-            newSelectedIDs = Collections.singleton(newSelectedEndpointID);
-        }
-
-        endpoint.selectedEndpointsChanged(newSelectedIDs);
+        endpoint.selectedEndpointsChanged(newSelectedEndpoints);
+        propagateJSONObject(jsonObject);
     }
 
     /**
-     * {@inheritDoc}
+     * Propagates the specified JSON object to all proxies (i.e. octo endpoints) of
+     * {@link #endpoint} to all remote bridges.
+     *
+     * @param jsonObject the JSON object to propagate.
      */
-    @Override
-    protected void onSelectedEndpointsChangedEvent(
-        Object src,
-        JSONObject jsonObject)
+    private void propagateJSONObject(JSONObject jsonObject)
     {
-        // Find the new pinned endpoint.
-        Object o = jsonObject.get("selectedEndpoints");
-        if (!(o instanceof JSONArray))
+        Conference conference = getConference();
+        if (conference == null || conference.isExpired())
         {
-            logger.warn("Received invalid or unexpected JSON: " + jsonObject);
+            logger.warn(
+                "Unable to propagate a JSON object, the conference is null or expired.");
             return;
         }
 
-        JSONArray jsonArray = (JSONArray) o;
-        Set<String> newSelectedEndpoints = new HashSet<>();
-        for (Object endpointId : jsonArray)
-        {
-            if (endpointId != null && endpointId instanceof String)
-            {
-                newSelectedEndpoints.add((String)endpointId);
-            }
-        }
+        jsonObject.put(PROP_TARGET_OCTO_ENDPOINT_ID, endpoint.getID());
```

Comment:
if this is propagating to octo, why is it in `EndpointMessageTransport` and not `OctoEndpointMessageTransport`?

Reply:
same answer as above